// docz-root/server/controllers/documentController.js
const Document = require("../models/Document");
const User = require("../models/User");
const ipfsService = require("../services/ipfsService");
const blockchainService = require("../services/blockchainService");
const logger = require("../utils/logger");
const { v4: uuidv4 } = require("uuid");
const crypto = require("crypto");

const DOCUMENT_TYPES = {
  property_deed: "Property Deed",
  affidavit: "Affidavit",
  certificate: "Certificate",
  agreement: "Agreement",
  court_order: "Court Order",
  identification: "Identification",
  contract: "Contract",
  will: "Will",
  power_of_attorney: "Power of Attorney",
  title_deed: "Title Deed",
  marriage_certificate: "Marriage Certificate",
  birth_certificate: "Birth Certificate",
  educational_certificate: "Educational Certificate",
  medical_record: "Medical Record",
  business_license: "Business License",
  tax_document: "Tax Document",
  insurance: "Insurance",
  other: "Other"
};

const uploadDocument = async (req, res) => {
  try {
    const { walletAddress } = req.user;
    const { title, documentType, description, tags } = req.body;

    let fileBuffer;
    let filename;
    let mimeType;

    if (req.file) {
      fileBuffer = req.file.buffer;
      filename = req.file.originalname;
      mimeType = req.file.mimetype;
    } else if (req.body.fileData) {
      fileBuffer = Buffer.from(req.body.fileData, "base64");
      filename = req.body.fileName || req.body.title || "document";
      mimeType = req.body.mimeType || "application/octet-stream";
    } else {
      return res.status(400).json({ error: "No file uploaded" });
    }

    if (!title || !documentType) {
      return res.status(400).json({ error: "Title and document type are required" });
    }

    const fileHash = crypto.createHash("sha256").update(fileBuffer).digest("hex");
    logger.info(`File hash: ${fileHash}`);

    const ipfsResult = await ipfsService.uploadFile(fileBuffer, filename, mimeType);
    const ipfsHash = ipfsResult.hash;

    const metadata = {
      title,
      documentType,
      description: description || "",
      originalHash: fileHash,
      filename,
      mimeType,
      uploadedAt: new Date().toISOString(),
      uploadedBy: walletAddress,
      fileSize: fileBuffer.length
    };

    const metadataResult = await ipfsService.uploadJSON(metadata);
    const metadataHash = metadataResult.hash;

    const uploadSalt = crypto.randomBytes(16).toString("hex");
    const combinedHash = crypto.createHash("sha256")
      .update(JSON.stringify({ fileHash, ipfsHash, metadataHash, uploadSalt }))
      .digest("hex");

    let blockchainResult;
    try {
      if (blockchainService.isInitialized) {
        blockchainResult = await blockchainService.storeDocument(
          ipfsHash,
          "0x" + combinedHash,
          documentType,
          description || ""
        );
      } else {
        blockchainResult = blockchainService.simulateTransaction({ type: "storeDocument" });
      }
    } catch (bcError) {
      logger.warn("Blockchain transaction failed, using simulation:", bcError.message);
      blockchainResult = blockchainService.simulateTransaction({ type: "storeDocument" });
    }

    const document = new Document({
      owner: walletAddress,
      documentHash: fileHash,
      ipfsHash,
      filename,
      mimeType,
      size: fileBuffer.length,
      documentType,
      title,
      description,
      blockchainTxHash: blockchainResult.txHash,
      blockchainDocHash: "0x" + combinedHash,
      status: "pending",
      tags: tags ? tags.split(",").map(t => t.trim()) : [],
      auditLog: [{
        action: "uploaded",
        performedBy: walletAddress,
        timestamp: new Date(),
        details: `Document "${title}" uploaded successfully`,
        ipAddress: req.ip,
        userAgent: req.get("user-agent")
      }]
    });

    await document.save();

    try {
      document.status = "verified";
      document.verifiedAt = new Date();
      await document.save();

      const emitNotification = req.app.get("emitNotification");
      if (emitNotification) {
        emitNotification(walletAddress, {
          type: "success",
          title: "Document Uploaded",
          message: `"${title}" has been secured on the blockchain`,
          documentId: document._id
        });
      }
    } catch (notifyError) {
      logger.warn("Notification error:", notifyError.message);
    }

    logger.info(`Document uploaded: ${document._id} by ${walletAddress}`);

    res.status(201).json({
      success: true,
      document: {
        id: document._id,
        title: document.title,
        documentType: document.documentType,
        documentTypeName: DOCUMENT_TYPES[document.documentType] || document.documentType,
        filename: document.filename,
        size: document.size,
        documentHash: document.documentHash,
        ipfsHash: document.ipfsHash,
        blockchainTxHash: document.blockchainTxHash,
        status: document.status,
        createdAt: document.createdAt
      }
    });
  } catch (error) {
    logger.error("Upload document error:", error);
    res.status(500).json({ error: "Failed to upload document" });
  }
};

const getDocuments = async (req, res) => {
  try {
    const { walletAddress, role } = req.user;
    const { type, status, search, page = 1, limit = 12 } = req.query;

    const query = {};

    if (role === "citizen" || role === "lawyer") {
      query.$or = [
        { owner: walletAddress.toLowerCase() },
        { "accessList.grantedTo": walletAddress.toLowerCase(), "accessList.active": true }
      ];
    }

    if (type) {
      query.documentType = type;
    }

    if (status) {
      query.status = status;
    }

    if (search) {
      query.$or = [
        { title: { $regex: search, $options: "i" } },
        { description: { $regex: search, $options: "i" } },
        { tags: { $regex: search, $options: "i" } }
      ];
    }

    const skip = (parseInt(page) - 1) * parseInt(limit);

    const [documents, total] = await Promise.all([
      Document.find(query)
        .sort({ createdAt: -1 })
        .skip(skip)
        .limit(parseInt(limit))
        .select("-auditLog"),
      Document.countDocuments(query)
    ]);

    const formattedDocs = documents.map(doc => ({
      id: doc._id,
      title: doc.title,
      documentType: doc.documentType,
      documentTypeName: DOCUMENT_TYPES[doc.documentType] || doc.documentType,
      filename: doc.filename,
      size: doc.size,
      documentHash: doc.documentHash,
      ipfsHash: doc.ipfsHash,
      blockchainTxHash: doc.blockchainTxHash,
      status: doc.status,
      isOwner: doc.owner.toLowerCase() === walletAddress.toLowerCase(),
      owner: doc.owner,
      tags: doc.tags,
      createdAt: doc.createdAt,
      verifiedAt: doc.verifiedAt
    }));

    res.json({
      documents: formattedDocs,
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total,
        pages: Math.ceil(total / parseInt(limit))
      }
    });
  } catch (error) {
    logger.error("Get documents error:", error);
    res.status(500).json({ error: "Failed to fetch documents" });
  }
};

const getDocumentById = async (req, res) => {
  try {
    const { walletAddress, role } = req.user;
    const { id } = req.params;

    const document = await Document.findById(id);

    if (!document) {
      return res.status(404).json({ error: "Document not found" });
    }

    if (!document.canAccess(walletAddress, role)) {
      return res.status(403).json({ error: "Access denied" });
    }

    const accessLevel = document.owner.toLowerCase() === walletAddress.toLowerCase()
      ? "owner"
      : document.accessList.find(a =>
          a.grantedTo.toLowerCase() === walletAddress.toLowerCase() && a.active
        )?.role || "view";

    const gatewayUrl = ipfsService.getGatewayUrl(document.ipfsHash);

    res.json({
      document: {
        id: document._id,
        title: document.title,
        documentType: document.documentType,
        documentTypeName: DOCUMENT_TYPES[document.documentType] || document.documentType,
        filename: document.filename,
        mimeType: document.mimeType,
        size: document.size,
        description: document.description,
        documentHash: document.documentHash,
        ipfsHash: document.ipfsHash,
        ipfsGatewayUrl: gatewayUrl,
        blockchainTxHash: document.blockchainTxHash,
        blockchainDocHash: document.blockchainDocHash,
        status: document.status,
        isOwner: document.owner.toLowerCase() === walletAddress.toLowerCase(),
        owner: document.owner,
        isRevoked: document.isRevoked,
        tags: document.tags,
        accessLevel,
        auditLog: accessLevel === "owner" || role === "admin" || role === "authority"
          ? document.auditLog
          : document.auditLog.filter(log => log.action !== "accessed"),
        accessList: accessLevel === "owner" || role === "admin"
          ? document.accessList
          : undefined,
        sharedLinks: accessLevel === "owner" || role === "admin"
          ? document.sharedLinks
          : undefined,
        createdAt: document.createdAt,
        verifiedAt: document.verifiedAt,
        updatedAt: document.updatedAt
      }
    });
  } catch (error) {
    logger.error("Get document error:", error);
    res.status(500).json({ error: "Failed to fetch document" });
  }
};

const verifyDocument = async (req, res) => {
  try {
    const { walletAddress } = req.user;
    const { documentId, file, documentHash } = req.body;

    let hashToVerify;
    let verificationMethod;
    let dbDocument = null;
    let chainDocHash = null;

    if (documentId) {
      dbDocument = await Document.findById(documentId);
      if (!dbDocument) {
        return res.status(404).json({ error: "Document not found" });
      }
      hashToVerify = dbDocument.documentHash;
      chainDocHash = dbDocument.blockchainDocHash || null;
      verificationMethod = "database";
    } else if (file) {
      const buffer = Buffer.from(file, "base64");
      hashToVerify = crypto.createHash("sha256").update(buffer).digest("hex");
      verificationMethod = "file_upload";
      dbDocument = await Document.findOne({ documentHash: hashToVerify })
        .sort({ createdAt: -1 });
      chainDocHash = dbDocument?.blockchainDocHash || null;
    } else if (documentHash) {
      verificationMethod = "hash_provided";
      const normalized = documentHash.replace(/^0x/i, "");
      const matchByFile = await Document.findOne({ documentHash: normalized })
        .sort({ createdAt: -1 });
      if (matchByFile) {
        dbDocument = matchByFile;
        hashToVerify = normalized;
        chainDocHash = matchByFile.blockchainDocHash || null;
      } else {
        const withPrefix = documentHash.startsWith("0x") ? documentHash : `0x${documentHash}`;
        const byChain = await Document.findOne({ blockchainDocHash: withPrefix })
          .sort({ createdAt: -1 });
        if (byChain) {
          dbDocument = byChain;
          hashToVerify = byChain.documentHash;
          chainDocHash = byChain.blockchainDocHash || null;
        } else {
          hashToVerify = normalized;
        }
      }
    } else {
      return res.status(400).json({ error: "Provide documentId, file, or documentHash" });
    }

    let blockchainVerified = false;
    let blockchainResult = null;

    try {
      if (blockchainService.isInitialized && chainDocHash) {
        blockchainVerified = await blockchainService.verifyDocument(chainDocHash);
        blockchainResult = { verified: blockchainVerified };
      }
    } catch (bcError) {
      logger.warn("Blockchain verification skipped:", bcError.message);
    }

    let status = "unknown";
    let details = {};

    if (dbDocument) {
      status = dbDocument.isRevoked ? "tampered" : "authentic";
      details = {
        storedHash: dbDocument.documentHash,
        storedOn: dbDocument.createdAt,
        ipfsHash: dbDocument.ipfsHash,
        owner: dbDocument.owner,
        documentType: DOCUMENT_TYPES[dbDocument.documentType] || dbDocument.documentType,
        title: dbDocument.title
      };
    }

    const isMatch = !!(dbDocument && dbDocument.documentHash === hashToVerify);
    const isValid = isMatch && !dbDocument?.isRevoked;

    if (documentId && dbDocument) {
      await dbDocument.addAuditEntry("verified", walletAddress, `Document verified as ${isValid ? "authentic" : "tampered"}`, req);
    }

    const io = req.app.get("io");
    if (io && dbDocument) {
      io.to(`user:${dbDocument.owner.toLowerCase()}`).emit("document_verified", {
        documentId: dbDocument._id,
        title: dbDocument.title,
        verifiedBy: walletAddress,
        result: isValid ? "authentic" : "tampered"
      });
    }

    res.json({
      result: !dbDocument ? "not_found" : (isValid ? "authentic" : "tampered"),
      isValid,
      isHashMatch: isMatch,
      verificationMethod,
      blockchainVerified,
      blockchainResult,
      details,
      verifiedAt: new Date().toISOString(),
      verifiedBy: walletAddress
    });
  } catch (error) {
    logger.error("Verify document error:", error);
    res.status(500).json({ error: "Failed to verify document" });
  }
};

const shareDocument = async (req, res) => {
  try {
    const { walletAddress } = req.user;
    const { id } = req.params;
    const { expiresIn } = req.body;

    const document = await Document.findById(id);

    if (!document) {
      return res.status(404).json({ error: "Document not found" });
    }

    if (document.owner.toLowerCase() !== walletAddress.toLowerCase()) {
      return res.status(403).json({ error: "Only owner can share documents" });
    }

    const linkId = uuidv4();
    const expiresAt = expiresIn
      ? new Date(Date.now() + expiresIn * 1000)
      : new Date(Date.now() + 7 * 24 * 60 * 60 * 1000);

    document.sharedLinks.push({
      linkId,
      createdAt: new Date(),
      expiresAt,
      accessCount: 0,
      active: true
    });

    await document.save();
    await document.addAuditEntry("shared", walletAddress, `Document shared via link`, req);

    const shareUrl = `${process.env.CLIENT_URL || "http://localhost:5173"}/verify/shared/${linkId}`;

    res.json({
      success: true,
      share: {
        linkId,
        shareUrl,
        expiresAt: expiresAt.toISOString(),
        qrCodeUrl: `https://api.qrserver.com/v1/create-qr-code/?size=200x200&data=${encodeURIComponent(shareUrl)}`
      }
    });
  } catch (error) {
    logger.error("Share document error:", error);
    res.status(500).json({ error: "Failed to share document" });
  }
};

const accessSharedLink = async (req, res) => {
  try {
    const { linkId } = req.params;

    const document = await Document.findOne({
      "sharedLinks.linkId": linkId,
      "sharedLinks.active": true
    });

    if (!document) {
      return res.status(404).json({ error: "Shared link not found or expired" });
    }

    const sharedLink = document.sharedLinks.find(l => l.linkId === linkId);

    if (!sharedLink || !sharedLink.active || sharedLink.expiresAt < new Date()) {
      return res.status(410).json({ error: "Shared link has expired" });
    }

    sharedLink.accessCount += 1;
    await document.save();

    await document.addAuditEntry(
      "accessed",
      req.user?.walletAddress || "anonymous",
      `Document accessed via shared link`,
      req
    );

    res.json({
      document: {
        id: document._id,
        title: document.title,
        documentType: DOCUMENT_TYPES[document.documentType] || document.documentType,
        description: document.description,
        owner: document.owner,
        status: document.isRevoked ? "revoked" : document.status,
        createdAt: document.createdAt,
        isRevoked: document.isRevoked
      },
      sharedBy: document.owner
    });
  } catch (error) {
    logger.error("Access shared link error:", error);
    res.status(500).json({ error: "Failed to access shared document" });
  }
};

const manageAccess = async (req, res) => {
  try {
    const { walletAddress } = req.user;
    const { id } = req.params;
    const { action, grantedTo, role, expiresIn } = req.body;

    const document = await Document.findById(id);

    if (!document) {
      return res.status(404).json({ error: "Document not found" });
    }

    if (document.owner.toLowerCase() !== walletAddress.toLowerCase()) {
      return res.status(403).json({ error: "Only owner can manage access" });
    }

    if (action === "grant") {
      if (!grantedTo || !role) {
        return res.status(400).json({ error: "grantedTo and role are required" });
      }

      const existingIndex = document.accessList.findIndex(
        a => a.grantedTo.toLowerCase() === grantedTo.toLowerCase()
      );

      const expiresAt = expiresIn
        ? new Date(Date.now() + expiresIn * 1000)
        : new Date(Date.now() + 30 * 24 * 60 * 60 * 1000);

      if (existingIndex >= 0) {
        document.accessList[existingIndex].active = true;
        document.accessList[existingIndex].role = role;
        document.accessList[existingIndex].expiresAt = expiresAt;
      } else {
        document.accessList.push({
          grantedTo: grantedTo.toLowerCase(),
          role,
          grantedAt: new Date(),
          expiresAt,
          active: true
        });
      }

      try {
        if (blockchainService.isInitialized) {
          const duration = expiresIn || 2592000;
          await blockchainService.grantAccess(
            document.blockchainDocHash,
            grantedTo,
            duration
          );
        }
      } catch (bcError) {
        logger.warn("Blockchain grant access error:", bcError.message);
      }

      await document.addAuditEntry(
        "access_granted",
        walletAddress,
        `Access granted to ${grantedTo} with role ${role}`,
        req
      );

      const emitNotification = req.app.get("emitNotification");
      if (emitNotification) {
        emitNotification(grantedTo.toLowerCase(), {
          type: "info",
          title: "Document Access Granted",
          message: `You have been granted ${role} access to "${document.title}"`,
          documentId: document._id
        });
      }
    } else if (action === "revoke") {
      if (!grantedTo) {
        return res.status(400).json({ error: "grantedTo is required" });
      }

      const accessIndex = document.accessList.findIndex(
        a => a.grantedTo.toLowerCase() === grantedTo.toLowerCase()
      );

      if (accessIndex >= 0) {
        document.accessList[accessIndex].active = false;
      }

      try {
        if (blockchainService.isInitialized) {
          await blockchainService.revokeAccess(
            document.blockchainDocHash,
            grantedTo
          );
        }
      } catch (bcError) {
        logger.warn("Blockchain revoke access error:", bcError.message);
      }

      await document.addAuditEntry(
        "access_revoked",
        walletAddress,
        `Access revoked from ${grantedTo}`,
        req
      );

      const emitNotification = req.app.get("emitNotification");
      if (emitNotification) {
        emitNotification(grantedTo.toLowerCase(), {
          type: "warning",
          title: "Document Access Revoked",
          message: `Your access to "${document.title}" has been revoked`,
          documentId: document._id
        });
      }
    } else {
      return res.status(400).json({ error: "Invalid action. Use 'grant' or 'revoke'" });
    }

    await document.save();

    res.json({
      success: true,
      accessList: document.accessList
    });
  } catch (error) {
    logger.error("Manage access error:", error);
    res.status(500).json({ error: "Failed to manage access" });
  }
};

const revokeDocument = async (req, res) => {
  try {
    const { walletAddress, role } = req.user;
    const { id } = req.params;

    const document = await Document.findById(id);

    if (!document) {
      return res.status(404).json({ error: "Document not found" });
    }

    if (document.owner.toLowerCase() !== walletAddress.toLowerCase() &&
        role !== "admin" && role !== "authority") {
      return res.status(403).json({ error: "Not authorized to revoke this document" });
    }

    document.isRevoked = true;
    document.status = "revoked";
    await document.save();

    try {
      if (blockchainService.isInitialized) {
        await blockchainService.revokeDocument(document.blockchainDocHash);
      }
    } catch (bcError) {
      logger.warn("Blockchain revoke error:", bcError.message);
    }

    await document.addAuditEntry("revoked", walletAddress, "Document revoked", req);

    const io = req.app.get("io");
    if (io) {
      document.accessList.forEach(access => {
        io.to(`user:${access.grantedTo.toLowerCase()}`).emit("document_revoked", {
          documentId: document._id,
          title: document.title,
          revokedBy: walletAddress
        });
      });
    }

    res.json({
      success: true,
      message: "Document revoked successfully"
    });
  } catch (error) {
    logger.error("Revoke document error:", error);
    res.status(500).json({ error: "Failed to revoke document" });
  }
};

const getAuditLog = async (req, res) => {
  try {
    const { walletAddress, role } = req.user;
    const { id } = req.params;

    const document = await Document.findById(id);

    if (!document) {
      return res.status(404).json({ error: "Document not found" });
    }

    if (!document.canAccess(walletAddress, role)) {
      return res.status(403).json({ error: "Access denied" });
    }

    res.json({
      auditLog: document.auditLog
    });
  } catch (error) {
    logger.error("Get audit log error:", error);
    res.status(500).json({ error: "Failed to fetch audit log" });
  }
};

const getStats = async (req, res) => {
  try {
    const { walletAddress, role } = req.user;

    const query = {};
    if (role === "citizen" || role === "lawyer") {
      query.$or = [
        { owner: walletAddress.toLowerCase() },
        { "accessList.grantedTo": walletAddress.toLowerCase(), "accessList.active": true }
      ];
    }

    const [total, verified, pending, revoked, recentActivity] = await Promise.all([
      Document.countDocuments(query),
      Document.countDocuments({ ...query, status: "verified" }),
      Document.countDocuments({ ...query, status: "pending" }),
      Document.countDocuments({ ...query, status: "revoked" }),
      Document.find(query)
        .sort({ updatedAt: -1 })
        .limit(5)
        .select("title status createdAt updatedAt")
    ]);

    res.json({
      stats: {
        total,
        verified,
        pending,
        revoked
      },
      recentActivity: recentActivity.map(doc => ({
        id: doc._id,
        title: doc.title,
        status: doc.status,
        createdAt: doc.createdAt,
        updatedAt: doc.updatedAt
      }))
    });
  } catch (error) {
    logger.error("Get stats error:", error);
    res.status(500).json({ error: "Failed to fetch stats" });
  }
};

const getDocumentTypes = async (req, res) => {
  res.json({
    types: Object.entries(DOCUMENT_TYPES).map(([value, label]) => ({
      value,
      label
    }))
  });
};

const deleteDocument = async (req, res) => {
  try {
    const { walletAddress, role } = req.user;
    const { id } = req.params;

    const document = await Document.findById(id);

    if (!document) {
      return res.status(404).json({ error: "Document not found" });
    }

    if (document.owner.toLowerCase() !== walletAddress.toLowerCase() && role !== "admin") {
      return res.status(403).json({ error: "Not authorized to delete this document" });
    }

    await Document.findByIdAndDelete(id);

    logger.info(`Document deleted: ${id} by ${walletAddress}`);

    res.json({
      success: true,
      message: "Document deleted successfully"
    });
  } catch (error) {
    logger.error("Delete document error:", error);
    res.status(500).json({ error: "Failed to delete document" });
  }
};

module.exports = {
  uploadDocument,
  getDocuments,
  getDocumentById,
  verifyDocument,
  shareDocument,
  accessSharedLink,
  manageAccess,
  revokeDocument,
  getAuditLog,
  getStats,
  getDocumentTypes,
  deleteDocument
};
