// docz-root/server/controllers/userController.js
const User = require("../models/User");
const bcrypt = require("bcryptjs");
const jwt = require("jsonwebtoken");
const logger = require("../utils/logger");

const JWT_SECRET = process.env.JWT_SECRET || "docz_super_secret_key_2024";

const registerOrLogin = async (req, res) => {
  try {
    const { walletAddress, signature, message } = req.body;

    if (!walletAddress) {
      return res.status(400).json({ error: "Wallet address is required" });
    }

    let user = await User.findOne({ walletAddress: walletAddress.toLowerCase() });

    if (!user) {
      user = new User({
        walletAddress: walletAddress.toLowerCase(),
        role: "citizen"
      });
      await user.save();
      logger.info(`New user registered: ${walletAddress}`);
    } else {
      user.lastLogin = new Date();
      await user.save();
    }

    const token = jwt.sign(
      {
        walletAddress: user.walletAddress,
        role: user.role,
        userId: user._id
      },
      JWT_SECRET,
      { expiresIn: "30d" }
    );

    res.json({
      success: true,
      user: user.toPublicJSON(),
      token
    });
  } catch (error) {
    logger.error("Register/login error:", error);
    res.status(500).json({ error: "Authentication failed" });
  }
};

const getProfile = async (req, res) => {
  try {
    const { walletAddress } = req.user;
    const user = await User.findOne({ walletAddress: walletAddress.toLowerCase() });

    if (!user) {
      return res.status(404).json({ error: "User not found" });
    }

    res.json({
      user: user.toPublicJSON()
    });
  } catch (error) {
    logger.error("Get profile error:", error);
    res.status(500).json({ error: "Failed to fetch profile" });
  }
};

const updateProfile = async (req, res) => {
  try {
    const { walletAddress } = req.user;
    const { name, email, phone, language, notificationPreferences } = req.body;

    const user = await User.findOne({ walletAddress: walletAddress.toLowerCase() });

    if (!user) {
      return res.status(404).json({ error: "User not found" });
    }

    if (name !== undefined) user.name = name;
    if (email !== undefined) user.email = email;
    if (phone !== undefined) user.phone = phone;
    if (language !== undefined) user.language = language;
    if (notificationPreferences !== undefined) {
      user.notificationPreferences = { ...user.notificationPreferences, ...notificationPreferences };
    }

    await user.save();

    res.json({
      success: true,
      user: user.toPublicJSON()
    });
  } catch (error) {
    logger.error("Update profile error:", error);
    res.status(500).json({ error: "Failed to update profile" });
  }
};

const getUserByWallet = async (req, res) => {
  try {
    const { wallet } = req.params;
    const user = await User.findOne({ walletAddress: wallet.toLowerCase() });

    if (!user) {
      return res.status(404).json({ error: "User not found" });
    }

    res.json({
      user: {
        walletAddress: user.walletAddress,
        role: user.role,
        name: user.name,
        organization: user.organization
      }
    });
  } catch (error) {
    logger.error("Get user by wallet error:", error);
    res.status(500).json({ error: "Failed to fetch user" });
  }
};

const verifyWallet = async (req, res) => {
  try {
    const { walletAddress } = req.body;

    if (!walletAddress) {
      return res.status(400).json({ error: "Wallet address is required" });
    }

    const user = await User.findOne({ walletAddress: walletAddress.toLowerCase() });

    res.json({
      registered: !!user,
      walletAddress: walletAddress.toLowerCase()
    });
  } catch (error) {
    logger.error("Verify wallet error:", error);
    res.status(500).json({ error: "Verification failed" });
  }
};

const listUsers = async (req, res) => {
  try {
    const { role, search, page = 1, limit = 20 } = req.query;

    if (req.user.role !== "admin" && req.user.role !== "authority") {
      return res.status(403).json({ error: "Not authorized" });
    }

    const query = {};

    if (role) {
      query.role = role;
    }

    if (search) {
      query.$or = [
        { name: { $regex: search, $options: "i" } },
        { walletAddress: { $regex: search, $options: "i" } },
        { organization: { $regex: search, $options: "i" } }
      ];
    }

    const skip = (parseInt(page) - 1) * parseInt(limit);

    const [users, total] = await Promise.all([
      User.find(query)
        .sort({ createdAt: -1 })
        .skip(skip)
        .limit(parseInt(limit))
        .select("-__v"),
      User.countDocuments(query)
    ]);

    res.json({
      users: users.map(u => u.toPublicJSON()),
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total,
        pages: Math.ceil(total / parseInt(limit))
      }
    });
  } catch (error) {
    logger.error("List users error:", error);
    res.status(500).json({ error: "Failed to fetch users" });
  }
};

const updateUserRole = async (req, res) => {
  try {
    if (req.user.role !== "admin") {
      return res.status(403).json({ error: "Not authorized" });
    }

    const { walletAddress } = req.params;
    const { role } = req.body;

    if (!["citizen", "lawyer", "authority", "admin"].includes(role)) {
      return res.status(400).json({ error: "Invalid role" });
    }

    const user = await User.findOne({ walletAddress: walletAddress.toLowerCase() });

    if (!user) {
      return res.status(404).json({ error: "User not found" });
    }

    user.role = role;
    await user.save();

    logger.info(`User ${walletAddress} role updated to ${role}`);

    res.json({
      success: true,
      user: user.toPublicJSON()
    });
  } catch (error) {
    logger.error("Update user role error:", error);
    res.status(500).json({ error: "Failed to update user role" });
  }
};

module.exports = {
  registerOrLogin,
  getProfile,
  updateProfile,
  getUserByWallet,
  verifyWallet,
  listUsers,
  updateUserRole
};
