// docz-root/api/services/blockchainService.js
const { ethers } = require("ethers");
const logger = require("../utils/logger");

class BlockchainService {
  constructor() {
    this.provider = null;
    this.signer = null;
    this.contract = null;
    this.contractAddress = process.env.CONTRACT_ADDRESS || "";
    this.isInitialized = false;
  }

  async initialize() {
    try {
      const network = process.env.NODE_ENV === "production" ? "polygonAmoy" : "polygonAmoy";
      const rpcUrl = process.env.POLYGON_AMOY_URL || "https://rpc-amoy.polygon.technology";

      this.provider = new ethers.JsonRpcProvider(rpcUrl);

      if (process.env.PRIVATE_KEY) {
        this.signer = new ethers.Wallet(process.env.PRIVATE_KEY, this.provider);
        logger.info(`Blockchain signer initialized with address: ${this.signer.address}`);
      }

      if (this.contractAddress) {
        await this.loadContract();
      }

      this.isInitialized = true;
      logger.info("Blockchain service initialized");
    } catch (error) {
      logger.error("Blockchain initialization error:", error.message);
      this.isInitialized = false;
    }
  }

  async loadContract() {
    const artifact = require("../artifacts/contracts/DocumentVault.sol/DocumentVault.json");
    this.contract = new ethers.Contract(
      this.contractAddress,
      artifact.abi,
      this.signer || this.provider
    );
    logger.info(`Contract loaded at: ${this.contractAddress}`);
  }

  getContract() {
    return this.contract;
  }

  getProvider() {
    return this.provider;
  }

  getSigner() {
    return this.signer;
  }

  async storeDocument(ipfsHash, documentHash, documentType, description) {
    if (!this.contract) {
      logger.error("Contract not initialized");
      throw new Error("Blockchain contract not initialized");
    }

    try {
      const tx = await this.contract.storeDocument(
        ipfsHash,
        documentHash,
        documentType,
        description,
        { gasLimit: 500000 }
      );

      logger.info(`Store document tx sent: ${tx.hash}`);
      const receipt = await tx.wait();

      const event = receipt.logs.find(log => {
        try {
          const parsed = this.contract.interface.parseLog(log);
          return parsed?.name === "DocumentStored";
        } catch {
          return false;
        }
      });

      if (event) {
        const parsed = this.contract.interface.parseLog(event);
        logger.info(`Document stored on blockchain: ${parsed.args.docHash}`);
      }

      return {
        txHash: receipt.hash,
        blockNumber: receipt.blockNumber,
        status: "success"
      };
    } catch (error) {
      logger.error("Store document error:", error.message);
      throw error;
    }
  }

  async verifyDocument(docHash) {
    if (!this.contract) {
      throw new Error("Blockchain contract not initialized");
    }

    try {
      let h = docHash;
      if (typeof h === "string" && h.length === 64 && !h.startsWith("0x")) {
        h = `0x${h}`;
      }
      const isValid = await this.contract.verifyDocument(h);
      return isValid;
    } catch (error) {
      logger.error("Verify document error:", error.message);
      throw error;
    }
  }

  async verifyDocumentWithHash(originalHash, claimedHash) {
    if (!this.contract) {
      throw new Error("Blockchain contract not initialized");
    }

    try {
      const isValid = await this.contract.verifyDocumentWithHash(originalHash, claimedHash);
      return isValid;
    } catch (error) {
      logger.error("Verify with hash error:", error.message);
      throw error;
    }
  }

  async getDocument(docHash, walletAddress) {
    if (!this.contract) {
      throw new Error("Blockchain contract not initialized");
    }

    try {
      const doc = await this.contract.getDocument(docHash);
      return {
        ipfsHash: doc[0],
        documentHash: doc[1],
        owner: doc[2],
        timestamp: new Date(Number(doc[3]) * 1000),
        documentType: doc[4],
        description: doc[5],
        isRevoked: doc[6]
      };
    } catch (error) {
      logger.error("Get document error:", error.message);
      throw error;
    }
  }

  async grantAccess(docHash, grantedTo, duration) {
    if (!this.contract) {
      throw new Error("Blockchain contract not initialized");
    }

    try {
      const tx = await this.contract.grantAccess(docHash, grantedTo, duration, { gasLimit: 200000 });
      const receipt = await tx.wait();
      return { txHash: receipt.hash, status: "success" };
    } catch (error) {
      logger.error("Grant access error:", error.message);
      throw error;
    }
  }

  async revokeAccess(docHash, revokedFrom) {
    if (!this.contract) {
      throw new Error("Blockchain contract not initialized");
    }

    try {
      const tx = await this.contract.revokeAccess(docHash, revokedFrom, { gasLimit: 200000 });
      const receipt = await tx.wait();
      return { txHash: receipt.hash, status: "success" };
    } catch (error) {
      logger.error("Revoke access error:", error.message);
      throw error;
    }
  }

  async revokeDocument(docHash) {
    if (!this.contract) {
      throw new Error("Blockchain contract not initialized");
    }

    try {
      const tx = await this.contract.revokeDocument(docHash, { gasLimit: 200000 });
      const receipt = await tx.wait();
      return { txHash: receipt.hash, status: "success" };
    } catch (error) {
      logger.error("Revoke document error:", error.message);
      throw error;
    }
  }

  async getUserDocuments(walletAddress) {
    if (!this.contract) {
      return [];
    }

    try {
      const docHashes = await this.contract.getUserDocuments(walletAddress);
      return docHashes;
    } catch (error) {
      logger.error("Get user documents error:", error.message);
      return [];
    }
  }

  async hasAccess(docHash, walletAddress) {
    if (!this.contract) {
      return false;
    }

    try {
      return await this.contract.hasAccess(docHash, walletAddress);
    } catch (error) {
      logger.error("Has access error:", error.message);
      return false;
    }
  }

  async getAccessList(docHash) {
    if (!this.contract) {
      return { addresses: [], expires: [], actives: [] };
    }

    try {
      const [addresses, expires, actives] = await this.contract.getAccessList(docHash);
      return { addresses, expires, actives };
    } catch (error) {
      logger.error("Get access list error:", error.message);
      return { addresses: [], expires: [], actives: [] };
    }
  }

  async requestAccess(docHash) {
    if (!this.contract) {
      throw new Error("Blockchain contract not initialized");
    }

    try {
      const tx = await this.contract.requestAccess(docHash, { gasLimit: 200000 });
      const receipt = await tx.wait();
      return { txHash: receipt.hash, status: "success" };
    } catch (error) {
      logger.error("Request access error:", error.message);
      throw error;
    }
  }

  async getNetworkInfo() {
    if (!this.provider) {
      return null;
    }

    try {
      const network = await this.provider.getNetwork();
      const blockNumber = await this.provider.getBlockNumber();
      const gasPrice = await this.provider.getFeeData();

      return {
        chainId: Number(network.chainId),
        name: network.name,
        blockNumber,
        gasPrice: gasPrice.gasPrice ? ethers.formatUnits(gasPrice.gasPrice, "gwei") : null
      };
    } catch (error) {
      logger.error("Get network info error:", error.message);
      return null;
    }
  }

  async getBalance(walletAddress) {
    if (!this.provider) {
      return null;
    }

    try {
      const balance = await this.provider.getBalance(walletAddress);
      return ethers.formatEther(balance);
    } catch (error) {
      logger.error("Get balance error:", error.message);
      return null;
    }
  }

  simulateTransaction(txData) {
    return {
      txHash: `0xsim_${Date.now()}_${Math.random().toString(36).substring(2, 15)}`,
      blockNumber: Math.floor(Math.random() * 10000000) + 50000000,
      status: "success",
      simulated: true,
      message: "Simulated transaction (blockchain not connected)"
    };
  }
}

module.exports = new BlockchainService();
