// docz-root/api/services/ipfsService.js
const axios = require("axios");
const FormData = require("form-data");
const logger = require("../utils/logger");

class IPFSService {
  constructor() {
    this.pinataApiKey = process.env.IPFS_API_KEY || "";
    this.pinataSecret = process.env.IPFS_API_SECRET || "";
    this.gatewayUrl = process.env.IPFS_GATEWAY_URL || "https://gateway.pinata.cloud/ipfs/";
    this.pinataEndpoint = "https://api.pinata.cloud/pinning/pinFileToIPFS";
    this.pinataJsonEndpoint = "https://api.pinata.cloud/pinning/pinJSONToIPFS";
  }

  getHeaders() {
    return {
      pinata_api_key: this.pinataApiKey,
      pinata_secret_api_key: this.pinataSecret
    };
  }

  async uploadFile(buffer, filename, mimeType) {
    try {
      if (!this.pinataApiKey || !this.pinataSecret) {
        logger.warn("Pinata credentials not configured, using local simulation");
        return this.simulateUpload(buffer, filename);
      }

      const formData = new FormData();
      formData.append("file", buffer, {
        filename,
        contentType: mimeType
      });

      const metadata = JSON.stringify({
        name: `DocZ+_${filename}_${Date.now()}`,
        keyvalues: {
          uploadedBy: "DocZ+",
          timestamp: new Date().toISOString()
        }
      });
      formData.append("pinataMetadata", metadata);

      const options = JSON.stringify({
        cidVersion: 1,
        wrapWithDirectory: false
      });
      formData.append("pinataOptions", options);

      const response = await axios.post(this.pinataEndpoint, formData, {
        headers: {
          ...formData.getHeaders(),
          ...this.getHeaders()
        },
        maxBodyLength: "Infinity",
        maxContentLength: "Infinity"
      });

      logger.info(`File uploaded to IPFS: ${response.data.IpfsHash}`);
      return {
        hash: response.data.IpfsHash,
        pinSize: response.data.PinSize,
        timestamp: new Date().toISOString()
      };
    } catch (error) {
      logger.error("IPFS upload error:", error.message);
      return this.simulateUpload(buffer, filename);
    }
  }

  async uploadJSON(data) {
    try {
      if (!this.pinataApiKey || !this.pinataSecret) {
        return this.simulateJSONUpload(data);
      }

      const body = {
        pinataContent: data,
        pinataMetadata: {
          name: `DocZ+_metadata_${Date.now()}`,
          keyvalues: {
            uploadedBy: "DocZ+",
            timestamp: new Date().toISOString()
          }
        },
        pinataOptions: {
          cidVersion: 1
        }
      };

      const response = await axios.post(this.pinataJsonEndpoint, body, {
        headers: this.getHeaders()
      });

      logger.info(`JSON uploaded to IPFS: ${response.data.IpfsHash}`);
      return {
        hash: response.data.IpfsHash,
        timestamp: new Date().toISOString()
      };
    } catch (error) {
      logger.error("IPFS JSON upload error:", error.message);
      return this.simulateJSONUpload(data);
    }
  }

  async pinHash(hash) {
    try {
      const response = await axios.post(
        "https://api.pinata.cloud/pinning/pinByHash",
        {
          hashToPin: hash,
          pinataMetadata: {
            name: `DocZ+_pin_${hash}`,
            keyvalues: {
              pinnedBy: "DocZ+",
              timestamp: new Date().toISOString()
            }
          }
        },
        { headers: this.getHeaders() }
      );
      return response.data;
    } catch (error) {
      logger.error("IPFS pin error:", error.message);
      return null;
    }
  }

  getGatewayUrl(hash) {
    return `${this.gatewayUrl}${hash}`;
  }

  async unpin(hash) {
    try {
      const response = await axios.delete(
        `https://api.pinata.cloud/pinning/unpin/${hash}`,
        { headers: this.getHeaders() }
      );
      return true;
    } catch (error) {
      logger.error("IPFS unpin error:", error.message);
      return false;
    }
  }

  async getFile(hash) {
    try {
      const gatewayUrl = this.getGatewayUrl(hash);
      const response = await axios.get(gatewayUrl, {
        responseType: "arraybuffer"
      });
      return {
        data: Buffer.from(response.data),
        mimeType: response.headers["content-type"]
      };
    } catch (error) {
      logger.error("IPFS get file error:", error.message);
      return null;
    }
  }

  simulateUpload(buffer, filename) {
    const hash = `Qm${Date.now()}_${Math.random().toString(36).substring(2, 15)}`;
    logger.info(`Simulated IPFS upload: ${hash}`);
    return {
      hash,
      pinSize: buffer.length,
      timestamp: new Date().toISOString(),
      simulated: true
    };
  }

  simulateJSONUpload(data) {
    const hash = `QmJSON${Date.now()}_${Math.random().toString(36).substring(2, 15)}`;
    logger.info(`Simulated JSON upload: ${hash}`);
    return {
      hash,
      timestamp: new Date().toISOString(),
      simulated: true
    };
  }
}

module.exports = new IPFSService();
