// docz-root/server/server.js
/**
 * Alternate Express + Socket.io backend. Not used by README / Vite default.
 * Primary local API: ../api/local-server.js (same Mongo models concept, different server).
 */
require("dotenv").config();
const express = require("express");
const cors = require("cors");
const helmet = require("helmet");
const rateLimit = require("express-rate-limit");
const http = require("http");
const { Server } = require("socket.io");
const path = require("path");

const connectDB = require("./config/db");
const logger = require("./utils/logger");

const documentRoutes = require("./routes/documentRoutes");
const userRoutes = require("./routes/userRoutes");

const app = express();
const server = http.createServer(app);

const io = new Server(server, {
  cors: {
    origin: process.env.CLIENT_URL || "http://localhost:5173",
    methods: ["GET", "POST"]
  }
});

connectDB();

const blockchainService = require("./services/blockchainService");
blockchainService.initialize().catch(err => {
  logger.warn("Blockchain service initialization failed (will use simulation):", err.message);
});

app.use(helmet({ contentSecurityPolicy: false }));
app.use(cors({
  origin: process.env.CLIENT_URL || "http://localhost:5173",
  credentials: true
}));
app.use(express.json({ limit: "50mb" }));
app.use(express.urlencoded({ extended: true, limit: "50mb" }));

const limiter = rateLimit({
  windowMs: 15 * 60 * 1000,
  max: 200,
  message: { error: "Too many requests, please try again later." }
});
app.use("/api/", limiter);

app.use("/api/documents", documentRoutes);
app.use("/api/users", userRoutes);

app.get("/api/health", (req, res) => {
  res.json({ status: "ok", timestamp: new Date().toISOString() });
});

app.use((err, req, res, next) => {
  logger.error(err.stack);
  res.status(err.status || 500).json({
    error: err.message || "Internal Server Error"
  });
});

app.use((req, res) => {
  res.status(404).json({ error: "Route not found" });
});

io.on("connection", (socket) => {
  logger.info(`Client connected: ${socket.id}`);

  socket.on("join_user", (walletAddress) => {
    socket.join(`user:${walletAddress.toLowerCase()}`);
    logger.info(`User ${walletAddress} joined their room`);
  });

  socket.on("disconnect", () => {
    logger.info(`Client disconnected: ${socket.id}`);
  });
});

const emitNotification = (walletAddress, notification) => {
  io.to(`user:${walletAddress.toLowerCase()}`).emit("notification", notification);
};

app.set("emitNotification", emitNotification);
app.set("io", io);

const PORT = process.env.PORT || 5000;

server.listen(PORT, () => {
  logger.info(`DocZ+ Server running on port ${PORT}`);
  logger.info(`Environment: ${process.env.NODE_ENV || "development"}`);
});

module.exports = { app, server, io };
