// docz-root/api/index.js
const connectDB = require('./config/db');
const User = require('./models/User');
const Document = require('./models/Document');
const ipfsService = require('./services/ipfsService');
const blockchainService = require('./services/blockchainService');
const logger = require('./utils/logger');
const jwt = require('jsonwebtoken');
const multer = require('multer');
const cors = require('cors');
const crypto = require('crypto');

const JWT_SECRET = process.env.JWT_SECRET || 'docz_super_secret_key_2024';

// Initialize services
let dbConnected = false;

const initDB = async () => {
  if (!dbConnected) {
    await connectDB();
    await blockchainService.initialize();
    dbConnected = true;
  }
};

// Multer config for file uploads
const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: 100 * 1024 * 1024 }
});

// Auth middleware
const authMiddleware = async (req, res, next) => {
  try {
    const token = req.headers.authorization?.split(' ')[1];
    if (!token) return res.status(401).json({ error: 'No token provided' });

    const decoded = jwt.verify(token, JWT_SECRET);
    req.user = {
      walletAddress: decoded.walletAddress,
      role: decoded.role,
      userId: decoded.userId
    };
    next();
  } catch (error) {
    return res.status(401).json({ error: 'Invalid token' });
  }
};

// Helper function to parse request body
const parseBody = (req) => {
  return new Promise((resolve, reject) => {
    let body = '';
    req.on('data', chunk => body += chunk);
    req.on('end', () => {
      try {
        if (req.headers['content-type']?.includes('multipart/form-data')) {
          resolve({});
        } else {
          resolve(body ? JSON.parse(body) : {});
        }
      } catch (e) {
        resolve({});
      }
    });
    req.on('error', reject);
  });
};

module.exports = async (req, res) => {
  // CORS
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');

  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }

  await initDB();

  const { url } = req;
  const pathname = url.split('?')[0];

  try {
    // Health check
    if (pathname === '/api/health') {
      return res.status(200).json({ status: 'ok', timestamp: new Date().toISOString() });
    }

    // User auth
    if (pathname === '/api/users/auth' && req.method === 'POST') {
      const body = await parseBody(req);
      const { walletAddress } = body;

      if (!walletAddress) {
        return res.status(400).json({ error: 'Wallet address required' });
      }

      let user = await User.findOne({ walletAddress: walletAddress.toLowerCase() });
      if (!user) {
        user = new User({ walletAddress: walletAddress.toLowerCase(), role: 'citizen' });
        await user.save();
      }

      const token = jwt.sign(
        { walletAddress: user.walletAddress, role: user.role, userId: user._id },
        JWT_SECRET,
        { expiresIn: '30d' }
      );

      return res.status(200).json({
        success: true,
        user: user.toPublicJSON(),
        token
      });
    }

    // Get profile
    if (pathname === '/api/users/profile' && req.method === 'GET') {
      const token = req.headers.authorization?.split(' ')[1];
      if (!token) return res.status(401).json({ error: 'Unauthorized' });

      const decoded = jwt.verify(token, JWT_SECRET);
      const user = await User.findOne({ walletAddress: decoded.walletAddress.toLowerCase() });
      if (!user) return res.status(404).json({ error: 'User not found' });

      return res.status(200).json({ user: user.toPublicJSON() });
    }

    // Update profile
    if (pathname === '/api/users/profile' && req.method === 'PUT') {
      const token = req.headers.authorization?.split(' ')[1];
      if (!token) return res.status(401).json({ error: 'Unauthorized' });

      const decoded = jwt.verify(token, JWT_SECRET);
      const body = await parseBody(req);
      const user = await User.findOne({ walletAddress: decoded.walletAddress.toLowerCase() });

      if (body.name) user.name = body.name;
      if (body.email) user.email = body.email;
      if (body.language) user.language = body.language;

      await user.save();
      return res.status(200).json({ success: true, user: user.toPublicJSON() });
    }

    // Verify wallet
    if (pathname.startsWith('/api/users/verify/')) {
      const wallet = pathname.split('/').pop();
      const user = await User.findOne({ walletAddress: wallet.toLowerCase() });
      return res.status(200).json({ registered: !!user, walletAddress: wallet.toLowerCase() });
    }

    // Document types
    if (pathname === '/api/documents/types' && req.method === 'GET') {
      const types = [
        { value: 'property_deed', label: 'Property Deed' },
        { value: 'affidavit', label: 'Affidavit' },
        { value: 'certificate', label: 'Certificate' },
        { value: 'agreement', label: 'Agreement' },
        { value: 'court_order', label: 'Court Order' },
        { value: 'identification', label: 'Identification' },
        { value: 'contract', label: 'Contract' },
        { value: 'will', label: 'Will' },
        { value: 'power_of_attorney', label: 'Power of Attorney' },
        { value: 'title_deed', label: 'Title Deed' },
        { value: 'marriage_certificate', label: 'Marriage Certificate' },
        { value: 'birth_certificate', label: 'Birth Certificate' },
        { value: 'educational_certificate', label: 'Educational Certificate' },
        { value: 'medical_record', label: 'Medical Record' },
        { value: 'business_license', label: 'Business License' },
        { value: 'tax_document', label: 'Tax Document' },
        { value: 'insurance', label: 'Insurance' },
        { value: 'other', label: 'Other' }
      ];
      return res.status(200).json({ types });
    }

    // Stats
    if (pathname === '/api/documents/stats' && req.method === 'GET') {
      const token = req.headers.authorization?.split(' ')[1];
      if (!token) return res.status(401).json({ error: 'Unauthorized' });

      const decoded = jwt.verify(token, JWT_SECRET);
      const wallet = decoded.walletAddress.toLowerCase();

      const query = {
        $or: [
          { owner: wallet },
          { 'accessList.grantedTo': wallet, 'accessList.active': true }
        ]
      };

      const [total, verified, pending, revoked] = await Promise.all([
        Document.countDocuments(query),
        Document.countDocuments({ ...query, status: 'verified' }),
        Document.countDocuments({ ...query, status: 'pending' }),
        Document.countDocuments({ ...query, status: 'revoked' })
      ]);

      return res.status(200).json({
        stats: { total, verified, pending, revoked },
        recentActivity: []
      });
    }

    // Get documents
    if (pathname === '/api/documents' && req.method === 'GET') {
      const token = req.headers.authorization?.split(' ')[1];
      if (!token) return res.status(401).json({ error: 'Unauthorized' });

      const decoded = jwt.verify(token, JWT_SECRET);
      const wallet = decoded.walletAddress.toLowerCase();
      const page = parseInt(req.query.page) || 1;
      const limit = parseInt(req.query.limit) || 12;
      const skip = (page - 1) * limit;

      const query = {
        $or: [
          { owner: wallet },
          { 'accessList.grantedTo': wallet, 'accessList.active': true }
        ]
      };

      if (req.query.search) {
        query.title = { $regex: req.query.search, $options: 'i' };
      }

      if (req.query.status) {
        query.status = req.query.status;
      }

      const [documents, total] = await Promise.all([
        Document.find(query).sort({ createdAt: -1 }).skip(skip).limit(limit).select('-auditLog'),
        Document.countDocuments(query)
      ]);

      const formattedDocs = documents.map(doc => ({
        id: doc._id,
        title: doc.title,
        documentType: doc.documentType,
        documentTypeName: doc.documentType.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase()),
        filename: doc.filename,
        size: doc.size,
        documentHash: doc.documentHash,
        ipfsHash: doc.ipfsHash,
        blockchainTxHash: doc.blockchainTxHash,
        status: doc.status,
        isOwner: doc.owner.toLowerCase() === wallet,
        owner: doc.owner,
        createdAt: doc.createdAt
      }));

      return res.status(200).json({
        documents: formattedDocs,
        pagination: { page, limit, total, pages: Math.ceil(total / limit) }
      });
    }

    // Get document by ID
    if (pathname.match(/^\/api\/documents\/[^/]+$/) && req.method === 'GET') {
      const token = req.headers.authorization?.split(' ')[1];
      if (!token) return res.status(401).json({ error: 'Unauthorized' });

      const decoded = jwt.verify(token, JWT_SECRET);
      const wallet = decoded.walletAddress.toLowerCase();
      const id = pathname.split('/')[3];

      const doc = await Document.findById(id);
      if (!doc) return res.status(404).json({ error: 'Document not found' });

      const canAccess = doc.owner.toLowerCase() === wallet ||
        decoded.role === 'admin' || decoded.role === 'authority' ||
        doc.accessList.some(a => a.grantedTo.toLowerCase() === wallet && a.active);

      if (!canAccess) return res.status(403).json({ error: 'Access denied' });

      return res.status(200).json({
        document: {
          id: doc._id,
          title: doc.title,
          documentType: doc.documentType,
          documentTypeName: doc.documentType.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase()),
          filename: doc.filename,
          size: doc.size,
          description: doc.description,
          documentHash: doc.documentHash,
          ipfsHash: doc.ipfsHash,
          blockchainTxHash: doc.blockchainTxHash,
          blockchainDocHash: doc.blockchainDocHash,
          status: doc.status,
          isOwner: doc.owner.toLowerCase() === wallet,
          owner: doc.owner,
          isRevoked: doc.isRevoked,
          tags: doc.tags,
          accessList: doc.accessList,
          auditLog: doc.auditLog,
          sharedLinks: doc.sharedLinks,
          createdAt: doc.createdAt,
          updatedAt: doc.updatedAt
        }
      });
    }

    // Upload document
    if (pathname === '/api/documents/upload' && req.method === 'POST') {
      const token = req.headers.authorization?.split(' ')[1];
      if (!token) return res.status(401).json({ error: 'Unauthorized' });

      const decoded = jwt.verify(token, JWT_SECRET);
      const wallet = decoded.walletAddress;

      // Parse multipart form data
      const body = await parseBody(req);
      const contentType = req.headers['content-type'] || '';

      // For serverless, we'll accept JSON with base64 file
      const { title, documentType, description, tags, fileData, fileName, mimeType } = body;

      if (!title || !documentType) {
        return res.status(400).json({ error: 'Title and document type required' });
      }

      let buffer, filename;
      if (fileData) {
        buffer = Buffer.from(fileData, 'base64');
        filename = fileName || title;
      } else {
        return res.status(400).json({ error: 'No file provided' });
      }

      const fileHash = crypto.createHash('sha256').update(buffer).digest('hex');
      const ipfsResult = await ipfsService.uploadFile(buffer, filename, mimeType || 'application/octet-stream');
      const ipfsHash = ipfsResult.hash;

      const uploadSalt = crypto.randomBytes(16).toString('hex');
      const combinedHash = crypto.createHash('sha256')
        .update(JSON.stringify({ fileHash, ipfsHash, uploadSalt }))
        .digest('hex');

      let blockchainResult;
      try {
        if (blockchainService.isInitialized) {
          blockchainResult = await blockchainService.storeDocument(
            ipfsHash, '0x' + combinedHash, documentType, description || ''
          );
        } else {
          blockchainResult = blockchainService.simulateTransaction({ type: 'storeDocument' });
        }
      } catch (bcError) {
        blockchainResult = blockchainService.simulateTransaction({ type: 'storeDocument' });
      }

      const document = new Document({
        owner: wallet,
        documentHash: fileHash,
        ipfsHash,
        filename,
        mimeType: mimeType || 'application/octet-stream',
        size: buffer.length,
        documentType,
        title,
        description,
        blockchainTxHash: blockchainResult.txHash,
        blockchainDocHash: '0x' + combinedHash,
        status: 'verified',
        tags: tags ? tags.split(',').map(t => t.trim()) : [],
        auditLog: [{
          action: 'uploaded',
          performedBy: wallet,
          timestamp: new Date(),
          details: `Document "${title}" uploaded successfully`
        }]
      });

      await document.save();

      return res.status(201).json({
        success: true,
        document: {
          id: document._id,
          title: document.title,
          documentType: document.documentType,
          documentTypeName: document.documentType.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase()),
          filename: document.filename,
          size: document.size,
          documentHash: document.documentHash,
          ipfsHash: document.ipfsHash,
          blockchainTxHash: document.blockchainTxHash,
          status: document.status,
          createdAt: document.createdAt
        }
      });
    }

    // Verify document
    if (pathname === '/api/documents/verify' && req.method === 'POST') {
      const body = await parseBody(req);
      const { documentId, documentHash, fileData } = body;

      let hashToVerify;

      if (documentId) {
        const doc = await Document.findById(documentId);
        if (!doc) return res.status(404).json({ error: 'Document not found' });
        hashToVerify = doc.documentHash;
      } else if (fileData) {
        const buffer = Buffer.from(fileData, 'base64');
        hashToVerify = crypto.createHash('sha256').update(buffer).digest('hex');
      } else if (documentHash) {
        hashToVerify = documentHash;
      } else {
        return res.status(400).json({ error: 'Provide documentId, file, or hash' });
      }

      let dbDocument = null;
      if (documentId) {
        dbDocument = await Document.findById(documentId);
      }

      const isValid = dbDocument ? dbDocument.documentHash === hashToVerify && !dbDocument.isRevoked : false;
      const result = dbDocument ? (dbDocument.isRevoked ? 'revoked' : 'authentic') : 'not_found';

      return res.status(200).json({
        result,
        isValid,
        details: dbDocument ? {
          title: dbDocument.title,
          storedHash: dbDocument.documentHash,
          uploadedOn: dbDocument.createdAt,
          owner: dbDocument.owner,
          documentType: dbDocument.documentType.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase()),
          ipfsHash: dbDocument.ipfsHash
        } : null,
        verifiedAt: new Date().toISOString()
      });
    }

    // Share document
    if (pathname.match(/^\/api\/documents\/[^/]+\/share$/) && req.method === 'POST') {
      const token = req.headers.authorization?.split(' ')[1];
      if (!token) return res.status(401).json({ error: 'Unauthorized' });

      const decoded = jwt.verify(token, JWT_SECRET);
      const id = pathname.split('/')[3];
      const body = await parseBody(req);
      const { expiresIn = 604800 } = body;

      const doc = await Document.findById(id);
      if (!doc) return res.status(404).json({ error: 'Document not found' });
      if (doc.owner.toLowerCase() !== decoded.walletAddress.toLowerCase()) {
        return res.status(403).json({ error: 'Only owner can share' });
      }

      const { v4: uuidv4 } = require('uuid');
      const linkId = uuidv4();
      const expiresAt = new Date(Date.now() + expiresIn * 1000);

      doc.sharedLinks.push({ linkId, createdAt: new Date(), expiresAt, accessCount: 0, active: true });
      await doc.save();

      const shareUrl = `${process.env.CLIENT_URL || 'http://localhost:5173'}/verify/shared/${linkId}`;

      return res.status(200).json({
        success: true,
        share: {
          linkId,
          shareUrl,
          expiresAt: expiresAt.toISOString(),
          qrCodeUrl: `https://api.qrserver.com/v1/create-qr-code/?size=200x200&data=${encodeURIComponent(shareUrl)}`
        }
      });
    }

    // Access shared link
    if (pathname.match(/^\/api\/documents\/shared\/[^/]+$/) && req.method === 'GET') {
      const linkId = pathname.split('/')[4];
      const doc = await Document.findOne({ 'sharedLinks.linkId': linkId, 'sharedLinks.active': true });

      if (!doc) return res.status(404).json({ error: 'Shared link not found or expired' });

      const sharedLink = doc.sharedLinks.find(l => l.linkId === linkId);
      if (!sharedLink || sharedLink.expiresAt < new Date()) {
        return res.status(410).json({ error: 'Shared link has expired' });
      }

      sharedLink.accessCount += 1;
      await doc.save();

      return res.status(200).json({
        document: {
          id: doc._id,
          title: doc.title,
          documentType: doc.documentType.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase()),
          description: doc.description,
          owner: doc.owner,
          status: doc.isRevoked ? 'revoked' : doc.status,
          createdAt: doc.createdAt,
          isRevoked: doc.isRevoked
        },
        sharedBy: doc.owner
      });
    }

    // Manage access
    if (pathname.match(/^\/api\/documents\/[^/]+\/access$/) && req.method === 'POST') {
      const token = req.headers.authorization?.split(' ')[1];
      if (!token) return res.status(401).json({ error: 'Unauthorized' });

      const decoded = jwt.verify(token, JWT_SECRET);
      const id = pathname.split('/')[3];
      const body = await parseBody(req);
      const { action, grantedTo, role = 'view', expiresIn = 2592000 } = body;

      const doc = await Document.findById(id);
      if (!doc) return res.status(404).json({ error: 'Document not found' });
      if (doc.owner.toLowerCase() !== decoded.walletAddress.toLowerCase()) {
        return res.status(403).json({ error: 'Only owner can manage access' });
      }

      if (action === 'grant') {
        if (!grantedTo) return res.status(400).json({ error: 'Wallet address required' });

        const expiresAt = new Date(Date.now() + expiresIn * 1000);
        const existing = doc.accessList.findIndex(a => a.grantedTo.toLowerCase() === grantedTo.toLowerCase());

        if (existing >= 0) {
          doc.accessList[existing].active = true;
          doc.accessList[existing].role = role;
          doc.accessList[existing].expiresAt = expiresAt;
        } else {
          doc.accessList.push({
            grantedTo: grantedTo.toLowerCase(),
            role,
            grantedAt: new Date(),
            expiresAt,
            active: true
          });
        }
      } else if (action === 'revoke') {
        if (!grantedTo) return res.status(400).json({ error: 'Wallet address required' });

        const idx = doc.accessList.findIndex(a => a.grantedTo.toLowerCase() === grantedTo.toLowerCase());
        if (idx >= 0) {
          doc.accessList[idx].active = false;
        }
      }

      await doc.save();
      return res.status(200).json({ success: true, accessList: doc.accessList });
    }

    // Revoke document
    if (pathname.match(/^\/api\/documents\/[^/]+\/revoke$/) && req.method === 'POST') {
      const token = req.headers.authorization?.split(' ')[1];
      if (!token) return res.status(401).json({ error: 'Unauthorized' });

      const decoded = jwt.verify(token, JWT_SECRET);
      const id = pathname.split('/')[3];

      const doc = await Document.findById(id);
      if (!doc) return res.status(404).json({ error: 'Document not found' });
      if (doc.owner.toLowerCase() !== decoded.walletAddress.toLowerCase() &&
          decoded.role !== 'admin' && decoded.role !== 'authority') {
        return res.status(403).json({ error: 'Not authorized' });
      }

      doc.isRevoked = true;
      doc.status = 'revoked';
      await doc.save();

      return res.status(200).json({ success: true, message: 'Document revoked' });
    }

    // Delete document
    if (pathname.match(/^\/api\/documents\/[^/]+$/) && req.method === 'DELETE') {
      const token = req.headers.authorization?.split(' ')[1];
      if (!token) return res.status(401).json({ error: 'Unauthorized' });

      const decoded = jwt.verify(token, JWT_SECRET);
      const id = pathname.split('/')[3];

      const doc = await Document.findById(id);
      if (!doc) return res.status(404).json({ error: 'Document not found' });
      if (doc.owner.toLowerCase() !== decoded.walletAddress.toLowerCase() && decoded.role !== 'admin') {
        return res.status(403).json({ error: 'Not authorized' });
      }

      await Document.findByIdAndDelete(id);
      return res.status(200).json({ success: true, message: 'Document deleted' });
    }

    // 404
    return res.status(404).json({ error: 'Route not found' });

  } catch (error) {
    logger.error('API error:', error);
    return res.status(500).json({ error: error.message || 'Internal server error' });
  }
};
