// docz-root/api/local-server.js
/**
 * Default local backend for DocZ+ (port 5000). Loads poli/.env.
 * For Vercel, use api/index.js. The Express stack lives in ../server/ (optional).
 */
require('dotenv').config({ path: '../.env' });
const http = require('http');
const connectDB = require('./config/db');
const User = require('./models/User');
const Document = require('./models/Document');
const ipfsService = require('./services/ipfsService');
const blockchainService = require('./services/blockchainService');
const logger = require('./utils/logger');
const jwt = require('jsonwebtoken');
const crypto = require('crypto');
const { v4: uuidv4 } = require('uuid');

const PORT = process.env.PORT || 5000;
const HOST = process.env.HOST || '0.0.0.0';
const JWT_SECRET = process.env.JWT_SECRET || 'docz_super_secret_key_2024';

const DOCUMENT_TYPES = [
  'property_deed', 'affidavit', 'certificate', 'agreement', 'court_order',
  'identification', 'contract', 'will', 'power_of_attorney', 'title_deed',
  'marriage_certificate', 'birth_certificate', 'educational_certificate',
  'medical_record', 'business_license', 'tax_document', 'insurance', 'other'
];

// Initialize services
let dbConnected = false;

const initServices = async () => {
  if (!dbConnected) {
    await connectDB();
    try {
      await blockchainService.initialize();
    } catch (e) {
      console.log('Blockchain service running in simulation mode');
    }
    dbConnected = true;
  }
};

// Parse JSON body
const parseJSON = (body) => {
  try {
    return JSON.parse(body);
  } catch {
    return {};
  }
};

// Send JSON response
const sendJSON = (res, status, data) => {
  res.writeHead(status, { 'Content-Type': 'application/json' });
  res.end(JSON.stringify(data));
};

// Auth middleware
const auth = (headers) => {
  try {
    const token = headers.authorization?.split(' ')[1];
    if (!token) return null;
    return jwt.verify(token, JWT_SECRET);
  } catch {
    return null;
  }
};

// CORS headers
const corsHeaders = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type, Authorization'
};

// Handle request
const handleRequest = async (req, res) => {
  // CORS
  Object.entries(corsHeaders).forEach(([k, v]) => res.setHeader(k, v));
  if (req.method === 'OPTIONS') {
    return sendJSON(res, 200, {});
  }

  await initServices();

  const url = new URL(req.url, `http://localhost:${PORT}`);
  const pathname = url.pathname;

  try {
    // Health check
    if (pathname === '/api/health') {
      return sendJSON(res, 200, { status: 'ok', timestamp: new Date().toISOString() });
    }

    // User auth
    if (pathname === '/api/users/auth' && req.method === 'POST') {
      let body = '';
      req.on('data', chunk => body += chunk);
      req.on('end', async () => {
        const { walletAddress } = parseJSON(body);
        if (!walletAddress) return sendJSON(res, 400, { error: 'Wallet address required' });

        let user = await User.findOne({ walletAddress: walletAddress.toLowerCase() });
        if (!user) {
          user = new User({ walletAddress: walletAddress.toLowerCase(), role: 'citizen' });
          await user.save();
        }

        const token = jwt.sign(
          { walletAddress: user.walletAddress, role: user.role, userId: user._id.toString() },
          JWT_SECRET,
          { expiresIn: '30d' }
        );

        sendJSON(res, 200, { success: true, user: user.toPublicJSON(), token });
      });
      return;
    }

    // Get profile
    if (pathname === '/api/users/profile' && req.method === 'GET') {
      const user = auth(req.headers);
      if (!user) return sendJSON(res, 401, { error: 'Unauthorized' });

      const profile = await User.findOne({ walletAddress: user.walletAddress.toLowerCase() });
      if (!profile) return sendJSON(res, 404, { error: 'User not found' });

      return sendJSON(res, 200, { user: profile.toPublicJSON() });
    }

    // Update profile
    if (pathname === '/api/users/profile' && req.method === 'PUT') {
      const user = auth(req.headers);
      if (!user) return sendJSON(res, 401, { error: 'Unauthorized' });

      let body = '';
      req.on('data', chunk => body += chunk);
      req.on('end', async () => {
        const data = parseJSON(body);
        const profile = await User.findOne({ walletAddress: user.walletAddress.toLowerCase() });
        if (data.name) profile.name = data.name;
        if (data.email) profile.email = data.email;
        if (data.language) profile.language = data.language;
        await profile.save();
        sendJSON(res, 200, { success: true, user: profile.toPublicJSON() });
      });
      return;
    }

    // Verify wallet
    if (pathname.startsWith('/api/users/verify/')) {
      const wallet = pathname.split('/').pop();
      const user = await User.findOne({ walletAddress: wallet.toLowerCase() });
      return sendJSON(res, 200, { registered: !!user, walletAddress: wallet.toLowerCase() });
    }

    // Document types
    if (pathname === '/api/documents/types') {
      const types = DOCUMENT_TYPES.map(v => ({ value: v, label: v.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase()) }));
      return sendJSON(res, 200, { types });
    }

    // Stats
    if (pathname === '/api/documents/stats') {
      const user = auth(req.headers);
      if (!user) return sendJSON(res, 401, { error: 'Unauthorized' });

      const wallet = user.walletAddress.toLowerCase();
      const query = {
        $or: [{ owner: wallet }, { 'accessList.grantedTo': wallet, 'accessList.active': true }]
      };

      const [total, verified, pending, revoked] = await Promise.all([
        Document.countDocuments(query),
        Document.countDocuments({ ...query, status: 'verified' }),
        Document.countDocuments({ ...query, status: 'pending' }),
        Document.countDocuments({ ...query, status: 'revoked' })
      ]);

      return sendJSON(res, 200, { stats: { total, verified, pending, revoked }, recentActivity: [] });
    }

    // Get documents
    if (pathname === '/api/documents' && req.method === 'GET') {
      const user = auth(req.headers);
      if (!user) return sendJSON(res, 401, { error: 'Unauthorized' });

      const wallet = user.walletAddress.toLowerCase();
      const page = parseInt(url.searchParams.get('page')) || 1;
      const limit = parseInt(url.searchParams.get('limit')) || 12;
      const search = url.searchParams.get('search');
      const status = url.searchParams.get('status');

      const query = {
        $or: [{ owner: wallet }, { 'accessList.grantedTo': wallet, 'accessList.active': true }]
      };
      if (search) query.title = { $regex: search, $options: 'i' };
      if (status) query.status = status;

      const [docs, total] = await Promise.all([
        Document.find(query).sort({ createdAt: -1 }).skip((page - 1) * limit).limit(limit),
        Document.countDocuments(query)
      ]);

      const documents = docs.map(doc => ({
        id: doc._id,
        title: doc.title,
        documentType: doc.documentType,
        documentTypeName: doc.documentType.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase()),
        filename: doc.filename,
        size: doc.size,
        documentHash: doc.documentHash,
        ipfsHash: doc.ipfsHash,
        blockchainTxHash: doc.blockchainTxHash,
        status: doc.status,
        isOwner: doc.owner.toLowerCase() === wallet,
        owner: doc.owner,
        createdAt: doc.createdAt
      }));

      return sendJSON(res, 200, { documents, pagination: { page, limit, total, pages: Math.ceil(total / limit) } });
    }

    // Get document by ID
    if (pathname.match(/^\/api\/documents\/[^/]+$/) && req.method === 'GET') {
      const user = auth(req.headers);
      if (!user) return sendJSON(res, 401, { error: 'Unauthorized' });

      const id = pathname.split('/')[3];
      const doc = await Document.findById(id);
      if (!doc) return sendJSON(res, 404, { error: 'Document not found' });

      const wallet = user.walletAddress.toLowerCase();
      const canAccess = doc.owner.toLowerCase() === wallet || user.role === 'admin' ||
        doc.accessList.some(a => a.grantedTo.toLowerCase() === wallet && a.active);
      if (!canAccess) return sendJSON(res, 403, { error: 'Access denied' });

      return sendJSON(res, 200, {
        document: {
          id: doc._id, title: doc.title, documentType: doc.documentType,
          documentTypeName: doc.documentType.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase()),
          filename: doc.filename, size: doc.size, description: doc.description,
          documentHash: doc.documentHash, ipfsHash: doc.ipfsHash,
          blockchainTxHash: doc.blockchainTxHash, blockchainDocHash: doc.blockchainDocHash,
          status: doc.status, isOwner: doc.owner.toLowerCase() === wallet,
          owner: doc.owner, isRevoked: doc.isRevoked, tags: doc.tags,
          accessList: doc.accessList, auditLog: doc.auditLog,
          sharedLinks: doc.sharedLinks, createdAt: doc.createdAt, updatedAt: doc.updatedAt
        }
      });
    }

    // Upload document
    if (pathname === '/api/documents/upload' && req.method === 'POST') {
      const user = auth(req.headers);
      if (!user) return sendJSON(res, 401, { error: 'Unauthorized' });

      let body = '';
      req.on('data', chunk => body += chunk);
      req.on('end', async () => {
        try {
          const data = parseJSON(body);
          const { title, documentType, description, tags, fileData, fileName, mimeType } = data;

          if (!title || !documentType || !fileData) {
            return sendJSON(res, 400, { error: 'Title, document type, and file required' });
          }

          const buffer = Buffer.from(fileData, 'base64');
          const fileHash = crypto.createHash('sha256').update(buffer).digest('hex');

          const ipfsResult = await ipfsService.uploadFile(buffer, fileName || title, mimeType || 'application/octet-stream');
          const ipfsHash = ipfsResult.hash;

          const uploadSalt = crypto.randomBytes(16).toString('hex');
          const combinedHash = crypto.createHash('sha256')
            .update(JSON.stringify({ fileHash, ipfsHash, uploadSalt }))
            .digest('hex');

          let blockchainResult = { txHash: `0xsim_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`, simulated: true };
          try {
            if (blockchainService.isInitialized) {
              blockchainResult = await blockchainService.storeDocument(ipfsHash, '0x' + combinedHash, documentType, description || '');
            }
          } catch (e) {}

          const doc = new Document({
            owner: user.walletAddress, documentHash: fileHash, ipfsHash, filename: fileName || title,
            mimeType: mimeType || 'application/octet-stream', size: buffer.length,
            documentType, title, description,
            blockchainTxHash: blockchainResult.txHash,
            blockchainDocHash: '0x' + combinedHash,
            status: 'verified',
            tags: tags ? tags.split(',').map(t => t.trim()) : [],
            auditLog: [{ action: 'uploaded', performedBy: user.walletAddress, timestamp: new Date(), details: `Document "${title}" uploaded` }]
          });
          await doc.save();

          return sendJSON(res, 201, {
            success: true,
            document: {
              id: doc._id, title: doc.title, documentType: doc.documentType,
              documentTypeName: doc.documentType.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase()),
              filename: doc.filename, size: doc.size, documentHash: doc.documentHash,
              ipfsHash: doc.ipfsHash, blockchainTxHash: doc.blockchainTxHash,
              status: doc.status, createdAt: doc.createdAt
            }
          });
        } catch (err) {
          return sendJSON(res, 500, { error: err.message });
        }
      });
      return;
    }

    // Verify document
    if (pathname === '/api/documents/verify' && req.method === 'POST') {
      let body = '';
      req.on('data', chunk => body += chunk);
      req.on('end', async () => {
        const { documentId, documentHash, fileData } = parseJSON(body);
        let hashToVerify;

        if (documentId) {
          const doc = await Document.findById(documentId);
          if (!doc) return sendJSON(res, 404, { error: 'Document not found' });
          hashToVerify = doc.documentHash;
        } else if (fileData) {
          const buffer = Buffer.from(fileData, 'base64');
          hashToVerify = crypto.createHash('sha256').update(buffer).digest('hex');
        } else if (documentHash) {
          hashToVerify = documentHash;
        } else {
          return sendJSON(res, 400, { error: 'Provide documentId, file, or hash' });
        }

        const doc = documentId ? await Document.findById(documentId) : null;
        const isValid = doc ? doc.documentHash === hashToVerify && !doc.isRevoked : false;
        const result = doc ? (doc.isRevoked ? 'revoked' : 'authentic') : 'not_found';

        return sendJSON(res, 200, {
          result, isValid,
          details: doc ? {
            title: doc.title, storedHash: doc.documentHash, uploadedOn: doc.createdAt,
            owner: doc.owner, documentType: doc.documentType.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase()),
            ipfsHash: doc.ipfsHash
          } : null,
          verifiedAt: new Date().toISOString()
        });
      });
      return;
    }

    // Share document
    if (pathname.match(/^\/api\/documents\/[^/]+\/share$/) && req.method === 'POST') {
      const user = auth(req.headers);
      if (!user) return sendJSON(res, 401, { error: 'Unauthorized' });

      let body = '';
      req.on('data', chunk => body += chunk);
      req.on('end', async () => {
        const id = pathname.split('/')[3];
        const { expiresIn = 604800 } = parseJSON(body);

        const doc = await Document.findById(id);
        if (!doc) return sendJSON(res, 404, { error: 'Document not found' });
        if (doc.owner.toLowerCase() !== user.walletAddress.toLowerCase()) {
          return sendJSON(res, 403, { error: 'Only owner can share' });
        }

        const linkId = uuidv4();
        const expiresAt = new Date(Date.now() + expiresIn * 1000);
        doc.sharedLinks.push({ linkId, createdAt: new Date(), expiresAt, accessCount: 0, active: true });
        await doc.save();

        const shareUrl = `http://localhost:${PORT}/verify/shared/${linkId}`;
        return sendJSON(res, 200, {
          success: true,
          share: {
            linkId, shareUrl, expiresAt: expiresAt.toISOString(),
            qrCodeUrl: `https://api.qrserver.com/v1/create-qr-code/?size=200x200&data=${encodeURIComponent(shareUrl)}`
          }
        });
      });
      return;
    }

    // Access shared link
    if (pathname.match(/^\/api\/documents\/shared\/[^/]+$/) && req.method === 'GET') {
      const linkId = pathname.split('/')[4];
      const doc = await Document.findOne({ 'sharedLinks.linkId': linkId, 'sharedLinks.active': true });
      if (!doc) return sendJSON(res, 404, { error: 'Shared link not found or expired' });

      const link = doc.sharedLinks.find(l => l.linkId === linkId);
      if (!link || link.expiresAt < new Date()) return sendJSON(res, 410, { error: 'Shared link expired' });

      link.accessCount += 1;
      await doc.save();

      return sendJSON(res, 200, {
        document: {
          id: doc._id, title: doc.title,
          documentType: doc.documentType.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase()),
          description: doc.description, owner: doc.owner,
          status: doc.isRevoked ? 'revoked' : doc.status,
          createdAt: doc.createdAt, isRevoked: doc.isRevoked
        },
        sharedBy: doc.owner
      });
    }

    // Manage access
    if (pathname.match(/^\/api\/documents\/[^/]+\/access$/) && req.method === 'POST') {
      const user = auth(req.headers);
      if (!user) return sendJSON(res, 401, { error: 'Unauthorized' });

      let body = '';
      req.on('data', chunk => body += chunk);
      req.on('end', async () => {
        const id = pathname.split('/')[3];
        const { action, grantedTo, role = 'view', expiresIn = 2592000 } = parseJSON(body);

        const doc = await Document.findById(id);
        if (!doc) return sendJSON(res, 404, { error: 'Document not found' });
        if (doc.owner.toLowerCase() !== user.walletAddress.toLowerCase()) {
          return sendJSON(res, 403, { error: 'Only owner can manage access' });
        }

        if (action === 'grant') {
          if (!grantedTo) return sendJSON(res, 400, { error: 'Wallet address required' });
          const expiresAt = new Date(Date.now() + expiresIn * 1000);
          const existing = doc.accessList.findIndex(a => a.grantedTo.toLowerCase() === grantedTo.toLowerCase());
          if (existing >= 0) {
            doc.accessList[existing] = { grantedTo: grantedTo.toLowerCase(), role, grantedAt: new Date(), expiresAt, active: true };
          } else {
            doc.accessList.push({ grantedTo: grantedTo.toLowerCase(), role, grantedAt: new Date(), expiresAt, active: true });
          }
        } else if (action === 'revoke') {
          if (!grantedTo) return sendJSON(res, 400, { error: 'Wallet address required' });
          const idx = doc.accessList.findIndex(a => a.grantedTo.toLowerCase() === grantedTo.toLowerCase());
          if (idx >= 0) doc.accessList[idx].active = false;
        }

        await doc.save();
        return sendJSON(res, 200, { success: true, accessList: doc.accessList });
      });
      return;
    }

    // Revoke document
    if (pathname.match(/^\/api\/documents\/[^/]+\/revoke$/) && req.method === 'POST') {
      const user = auth(req.headers);
      if (!user) return sendJSON(res, 401, { error: 'Unauthorized' });

      const id = pathname.split('/')[3];
      const doc = await Document.findById(id);
      if (!doc) return sendJSON(res, 404, { error: 'Document not found' });
      if (doc.owner.toLowerCase() !== user.walletAddress.toLowerCase() && user.role !== 'admin') {
        return sendJSON(res, 403, { error: 'Not authorized' });
      }

      doc.isRevoked = true;
      doc.status = 'revoked';
      await doc.save();
      return sendJSON(res, 200, { success: true, message: 'Document revoked' });
    }

    // Delete document
    if (pathname.match(/^\/api\/documents\/[^/]+$/) && req.method === 'DELETE') {
      const user = auth(req.headers);
      if (!user) return sendJSON(res, 401, { error: 'Unauthorized' });

      const id = pathname.split('/')[3];
      const doc = await Document.findById(id);
      if (!doc) return sendJSON(res, 404, { error: 'Document not found' });
      if (doc.owner.toLowerCase() !== user.walletAddress.toLowerCase() && user.role !== 'admin') {
        return sendJSON(res, 403, { error: 'Not authorized' });
      }

      await Document.findByIdAndDelete(id);
      return sendJSON(res, 200, { success: true, message: 'Document deleted' });
    }

    // 404
    return sendJSON(res, 404, { error: 'Route not found' });

  } catch (err) {
    logger.error('API error:', err);
    return sendJSON(res, 500, { error: err.message || 'Internal server error' });
  }
};

// Start server
const server = http.createServer(handleRequest);

server.listen(PORT, HOST, async () => {
  console.log(`\n🚀 DocZ+ API Server on http://${HOST}:${PORT}`);
  console.log(`📊 Health: /api/health\n`);
  await initServices();
});

module.exports = server;
