// docz-root/server/models/Document.js
const mongoose = require("mongoose");

const documentSchema = new mongoose.Schema({
  owner: {
    type: String,
    required: true,
    lowercase: true,
    trim: true,
    index: true
  },
  documentHash: {
    type: String,
    required: true,
    index: true
  },
  ipfsHash: {
    type: String,
    required: true
  },
  filename: {
    type: String,
    required: true
  },
  mimeType: {
    type: String
  },
  size: {
    type: Number
  },
  documentType: {
    type: String,
    required: true,
    enum: [
      "property_deed",
      "affidavit",
      "certificate",
      "agreement",
      "court_order",
      "identification",
      "contract",
      "will",
      "power_of_attorney",
      "title_deed",
      "marriage_certificate",
      "birth_certificate",
      "educational_certificate",
      "medical_record",
      "business_license",
      "tax_document",
      "insurance",
      "other"
    ]
  },
  title: {
    type: String,
    required: true,
    trim: true
  },
  description: {
    type: String,
    trim: true
  },
  blockchainTxHash: {
    type: String
  },
  blockchainDocHash: {
    type: String
  },
  status: {
    type: String,
    enum: ["pending", "verified", "rejected", "revoked"],
    default: "pending"
  },
  isRevoked: {
    type: Boolean,
    default: false
  },
  verifiedAt: {
    type: Date
  },
  accessList: [{
    grantedTo: {
      type: String,
      lowercase: true
    },
    role: {
      type: String,
      enum: ["view", "verify", "full"]
    },
    grantedAt: {
      type: Date,
      default: Date.now
    },
    expiresAt: {
      type: Date
    },
    active: {
      type: Boolean,
      default: true
    }
  }],
  sharedLinks: [{
    linkId: {
      type: String
    },
    createdAt: {
      type: Date,
      default: Date.now
    },
    expiresAt: {
      type: Date
    },
    accessCount: {
      type: Number,
      default: 0
    },
    active: {
      type: Boolean,
      default: true
    }
  }],
  auditLog: [{
    action: {
      type: String,
      enum: ["uploaded", "verified", "shared", "accessed", "revoked", "access_granted", "access_revoked"]
    },
    performedBy: {
      type: String
    },
    timestamp: {
      type: Date,
      default: Date.now
    },
    details: {
      type: String
    },
    ipAddress: {
      type: String
    },
    userAgent: {
      type: String
    }
  }],
  tags: [{
    type: String,
    trim: true
  }],
  expiresAt: {
    type: Date
  }
}, {
  timestamps: true
});

documentSchema.index({ owner: 1, createdAt: -1 });
documentSchema.index({ documentType: 1 });
documentSchema.index({ status: 1 });
documentSchema.index({ "accessList.grantedTo": 1 });
documentSchema.index({ "sharedLinks.linkId": 1 });
documentSchema.index({ tags: 1 });

documentSchema.methods.addAuditEntry = async function(action, performedBy, details, req = null) {
  const entry = {
    action,
    performedBy,
    timestamp: new Date(),
    details
  };

  if (req) {
    entry.ipAddress = req.ip || req.connection?.remoteAddress;
    entry.userAgent = req.get("user-agent");
  }

  this.auditLog.push(entry);
  await this.save();
};

documentSchema.methods.canAccess = function(walletAddress, role = "citizen") {
  if (this.owner.toLowerCase() === walletAddress.toLowerCase()) return true;
  if (role === "authority" || role === "admin") return true;

  const access = this.accessList.find(a =>
    a.grantedTo.toLowerCase() === walletAddress.toLowerCase() &&
    a.active &&
    (!a.expiresAt || a.expiresAt > new Date())
  );

  return !!access;
};

module.exports = mongoose.model("Document", documentSchema);
