// docz-root/api/models/User.js
const mongoose = require("mongoose");

const userSchema = new mongoose.Schema({
  walletAddress: {
    type: String,
    required: true,
    unique: true,
    lowercase: true,
    trim: true
  },
  email: {
    type: String,
    sparse: true,
    lowercase: true,
    trim: true
  },
  phone: {
    type: String,
    sparse: true,
    trim: true
  },
  role: {
    type: String,
    enum: ["citizen", "lawyer", "authority", "admin"],
    default: "citizen"
  },
  name: {
    type: String,
    trim: true
  },
  organization: {
    type: String,
    trim: true
  },
  notificationPreferences: {
    email: { type: Boolean, default: true },
    inApp: { type: Boolean, default: true },
    push: { type: Boolean, default: false }
  },
  language: {
    type: String,
    enum: ["en", "hi", "kn"],
    default: "en"
  },
  createdAt: {
    type: Date,
    default: Date.now
  },
  lastLogin: {
    type: Date
  },
  isActive: {
    type: Boolean,
    default: true
  }
}, {
  timestamps: true
});

userSchema.index({ role: 1 });

userSchema.methods.toPublicJSON = function() {
  return {
    walletAddress: this.walletAddress,
    name: this.name,
    role: this.role,
    organization: this.organization,
    language: this.language,
    createdAt: this.createdAt
  };
};

module.exports = mongoose.model("User", userSchema);
